File: backend/internal/models/fim_known_hosts.go
// Copyright (c) 2026 OPS Platform Contributors.
// SPDX-License-Identifier: MIT

package models

import "time"

// FIMKnownHost 已知主机密钥
type FIMKnownHost struct {
	ID          uint   `json:"id" gorm:"primaryKey"`
	Hostname    string `json:"hostname" gorm:"size:255;not null;uniqueIndex:uk_host_port_keytype"`
	Port        int    `json:"port" gorm:"not null;default:22;uniqueIndex:uk_host_port_keytype"`

	KeyType           string `json:"key_type" gorm:"size:50;not null;uniqueIndex:uk_host_port_keytype"`
	PublicKey         string `json:"public_key" gorm:"type:text;not null"`
	FingerprintSHA256 string `json:"fingerprint_sha256" gorm:"size:64;not null"`

	ServerID    *uint  `json:"server_id"`
	Description string `json:"description" gorm:"size:500"`
	Tags        string `json:"tags" gorm:"type:json"` // JSON array

	VerificationStatus string     `json:"verification_status" gorm:"size:20;default:verified"`
	VerifiedBy         string     `json:"verified_by" gorm:"size:100"`
	VerifiedAt         *time.Time `json:"verified_at"`

	LastUsedAt *time.Time `json:"last_used_at"`
	UseCount   int        `json:"use_count" gorm:"default:0"`

	AddedBy   string    `json:"added_by" gorm:"size:100;not null"`
	AddedAt   time.Time `json:"added_at" gorm:"autoCreateTime"`
	UpdatedBy string    `json:"updated_by" gorm:"size:100"`
	UpdatedAt time.Time `json:"updated_at" gorm:"autoUpdateTime"`

	IsEnabled bool `json:"is_enabled" gorm:"default:true"`
}

func (FIMKnownHost) TableName() string {
	return "fim_known_hosts"
}

// FIMKnownHostsHistory 主机密钥变更历史
type FIMKnownHostsHistory struct {
	ID      uint   `json:"id" gorm:"primaryKey"`
	HostID  uint   `json:"host_id" gorm:"not null;index"`
	Action  string `json:"action" gorm:"size:20;not null"`

	OldKeyType     *string `json:"old_key_type" gorm:"size:50"`
	OldPublicKey   *string `json:"old_public_key" gorm:"type:text"`
	OldFingerprint *string `json:"old_fingerprint" gorm:"size:64"`

	NewKeyType     *string `json:"new_key_type" gorm:"size:50"`
	NewPublicKey   *string `json:"new_public_key" gorm:"type:text"`
	NewFingerprint *string `json:"new_fingerprint" gorm:"size:64"`

	OperatedBy string    `json:"operated_by" gorm:"size:100;not null"`
	OperatedAt time.Time `json:"operated_at" gorm:"autoCreateTime"`
	Reason     string    `json:"reason" gorm:"type:text"`
	IPAddress  string    `json:"ip_address" gorm:"size:45"`
}

func (FIMKnownHostsHistory) TableName() string {
	return "fim_known_hosts_history"
}

// FIMSSHConnectionLog SSH连接尝试日志
type FIMSSHConnectionLog struct {
	ID       uint   `json:"id" gorm:"primaryKey"`
	Hostname string `json:"hostname" gorm:"size:255;not null;index:idx_hostname_port"`
	Port     int    `json:"port" gorm:"not null;index:idx_hostname_port"`

	Result       string `json:"result" gorm:"size:20;not null;index"`
	ErrorMessage string `json:"error_message" gorm:"type:text"`

	PresentedKeyType     *string `json:"presented_key_type" gorm:"size:50"`
	PresentedFingerprint *string `json:"presented_fingerprint" gorm:"size:64"`
	ExpectedFingerprint  *string `json:"expected_fingerprint" gorm:"size:64"`

	AttemptedAt time.Time `json:"attempted_at" gorm:"autoCreateTime;index"`
	ServerID    *uint     `json:"server_id" gorm:"index"`
	PolicyID    *uint     `json:"policy_id"`
	SnapshotID  *uint     `json:"snapshot_id"`
}

func (FIMSSHConnectionLog) TableName() string {
	return "fim_ssh_connection_logs"
}

File: frontend/src/api/fim-known-hosts.ts
// Copyright (c) 2026 OPS Platform Contributors.
// SPDX-License-Identifier: MIT

import client from './client';

export interface KnownHost {
  id: number;
  hostname: string;
  port: number;
  key_type: string;
  public_key: string;
  fingerprint_sha256: string;
  server_id?: number;
  description?: string;
  tags?: string; // JSON string in database
  verification_status: string;
  verified_by?: string;
  verified_at?: string;
  added_by: string;
  added_at: string;
  last_used_at?: string;
  use_count: number;
  is_enabled: boolean;
}

export interface ConnectionLog {
  id: number;
  hostname: string;
  port: number;
  result: 'success' | 'key_not_found' | 'key_mismatch' | 'connection_failed';
  error_message?: string;
  presented_key_type?: string;
  presented_fingerprint?: string;
  expected_fingerprint?: string;
  attempted_at: string;
}

export interface HostHistory {
  id: number;
  action: 'added' | 'updated' | 'deleted' | 'key_changed';
  old_key_type?: string;
  old_fingerprint?: string;
  new_key_type?: string;
  new_fingerprint?: string;
  operated_by: string;
  operated_at: string;
  reason?: string;
}

export interface AddKnownHostRequest {
  hostname: string;
  port: number;
  key_type: string;
  public_key: string;
  server_id?: number;
  description?: string;
  tags?: string[];
}

export interface UpdateKnownHostRequest {
  description?: string;
  tags?: string[];
  is_enabled?: boolean;
  public_key?: string;
}

export const fimKnownHostsAPI = {
  // 获取已知主机列表
  getHosts: async (params?: { hostname?: string; status?: string; key_type?: string }): Promise<{ data: KnownHost[]; total: number }> => {
    const response: any = await client.get('/security/fim/known-hosts', { params });
    return response;
  },

  // 获取单个主机详情
  getHost: async (id: number): Promise<{ host: KnownHost; logs: ConnectionLog[]; history: HostHistory[] }> => {
    const response: any = await client.get(`/security/fim/known-hosts/${id}`);
    return response.data;
  },

  // 添加已知主机
  addHost: async (data: AddKnownHostRequest): Promise<KnownHost> => {
    const response: any = await client.post('/security/fim/known-hosts', data);
    return response.data;
  },

  // 更新已知主机
  updateHost: async (id: number, data: UpdateKnownHostRequest): Promise<KnownHost> => {
    const response: any = await client.put(`/security/fim/known-hosts/${id}`, data);
    return response.data;
  },

  // 删除已知主机
  deleteHost: async (id: number): Promise<{ success: boolean }> => {
    const response: any = await client.delete(`/security/fim/known-hosts/${id}`);
    return response.data;
  },

  // 批量添加主机
  batchAddHosts: async (hosts: AddKnownHostRequest[]): Promise<{ success: boolean; added: number; skipped: number; errors: string[] }> => {
    const response: any = await client.post('/security/fim/known-hosts/batch', { hosts });
    return response.data;
  },

  // 导入known_hosts文件
  importHosts: async (content: string): Promise<{ success: boolean; imported: number; skipped: number; errors: string[] }> => {
    const response: any = await client.post('/security/fim/known-hosts/import', { content });
    return response.data;
  },

  // 导出known_hosts文件
  exportHosts: async (): Promise<Blob> => {
    const response: any = await client.get('/security/fim/known-hosts/export', {
      responseType: 'blob',
    });
    return response.data;
  },

  // 获取连接日志
  getConnectionLogs: async (params?: { hostname?: string; result?: string; limit?: number }): Promise<{ data: ConnectionLog[]; total: number }> => {
    const response: any = await client.get('/security/fim/connection-logs', { params });
    return response.data;
  },
};

export default fimKnownHostsAPI;